Sources/Rego/IREvaluator.swift

```diff
@@ -2,6 +2,119 @@ import AST
 import Foundation
 import IR
 
+// Custom array type that supports IR.Local indexing
+internal struct Locals: Equatable, Sendable, Encodable {
+    private var storage: Array<AST.RegoValue?>
+
+    init() {
+        self.storage = []
+    }
+
+    init(_ elements: [AST.RegoValue?]) {
+        self.storage = elements
+    }
+
+    init(repeating value: AST.RegoValue?, count: Int) {
+        self.storage = Array(repeating: value, count: count)
+    }
+
+    init(accommodating locals: [IR.Local], minimumSize: Int = 0) {
+        let maxLocal = locals.max() ?? IR.Local(0)
+        let requiredSize = Int(maxLocal) + 1 + minimumSize
+        self.storage = Array(repeating: nil, count: requiredSize)
+    }
+
+    // For tests only - converts dictionary to Locals array
+    init(from dict: [Int: Any]) {
+        guard let maxKey = dict.keys.max() else {
+            self.storage = []
+            return
+        }
+
+        self.storage = Array(repeating: nil, count: maxKey + 1)
+        for (index, value) in dict {
+            self.storage[index] = Self.convertToRegoValue(value)
+        }
+    }
+
+    private static func convertToRegoValue(_ value: Any) -> AST.RegoValue {
+        switch value {
+        case let string as String:
+            return .string(string)
+        case let int as Int:
+            return .number(NSNumber(value: int))
+        case let bool as Bool:
+            return .boolean(bool)
+        case let array as [Any]:
+            return .array(array.map(convertToRegoValue))
+        case let dict as [String: Any]:
+            let regoDict = Dictionary(uniqueKeysWithValues: dict.map { (key, val) in
+                (AST.RegoValue.string(key), convertToRegoValue(val))
+            })
+            return .object(regoDict)
+        case let regoValue as AST.RegoValue:
+            return regoValue
+        default:
+            return .undefined
+        }
+    }
+
+    subscript(index: IR.Local) -> AST.RegoValue? {
+        get {
+            let intIndex = Int(index)
+            return intIndex < storage.count ? storage[intIndex] : nil
```

Comment:
⛏️ I know `IR.Local` is unsigned but should we double check the lower bound in case something elsewhere in the code base changes?

```
    guard intIndex >= 0, intIndex < storage.count else {
        return nil
    }
    return storage[intIndex]
```